Source/Details/ASCollectionLayoutState.m

```diff
@@ -41,21 +62,74 @@ - (instancetype)initWithElements:(ASElementMap *)elements layout:(ASLayout *)lay
     }
     
     attrs.frame = sublayout.frame;
-    [attrsMap setObject:attrs forKey:element];
+    [table setObject:attrs forKey:element];
   }
 
-  return [self initWithElements:elements contentSize:layout.size elementToLayoutArrtibutesMap:attrsMap];
+  return [self initWithContext:context contentSize:layout.size elementToLayoutAttributesTable:table];
 }
 
-- (instancetype)initWithElements:(ASElementMap *)elements contentSize:(CGSize)contentSize elementToLayoutArrtibutesMap:(NSMapTable<ASCollectionElement *,UICollectionViewLayoutAttributes *> *)attrsMap
+- (instancetype)initWithContext:(ASCollectionLayoutContext *)context contentSize:(CGSize)contentSize elementToLayoutAttributesTable:(NSMapTable<ASCollectionElement *,UICollectionViewLayoutAttributes *> *)table
 {
   self = [super init];
   if (self) {
-    _elements = elements;
+    _context = context;
     _contentSize = contentSize;
-    _elementToLayoutArrtibutesMap = attrsMap;
+    _contentRect = CGRectMake(0.0f, 0.0f, contentSize.width, contentSize.height);
+    _elementToLayoutAttributesTable = table;
+    _pageToLayoutAttributesTable = [ASPageTable pageTableWithLayoutAttributes:_elementToLayoutAttributesTable.objectEnumerator pageSize:context.viewportSize];
   }
   return self;
 }
 
+- (NSArray<UICollectionViewLayoutAttributes *> *)allLayoutAttributes
+{
+  return [_elementToLayoutAttributesTable.objectEnumerator allObjects];
+}
+
+- (NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect
+{
+  CGSize pageSize = _context.viewportSize;
+  NSPointerArray *pages = ASPageCoordinatesForPagesThatIntersectRect(rect, _contentSize, pageSize);
+  if (pages.count == 0) {
+    return nil;
+  }
+  
+  NSMutableArray<UICollectionViewLayoutAttributes *> *results = [NSMutableArray array];
```

Comment:
Let's make this a mutable set, since layout attributes could be span multiple pages and we want to dedupe (see previous comment).